spam/spam_data_augmentation_tutorial.py

```diff
@@ -0,0 +1,301 @@
+# The two main columns in the DataFrames are:
+# * **`text`**: Raw text content of the comment
+# * **`label`**: Whether the comment is `SPAM` (1), `HAM` (0), or `UNKNOWN/ABSTAIN` (-1)
+#
+# For more details, check out the labeling functions [tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/spam_tutorial.ipynb).
+
+# %%
+# %load_ext autoreload
+# %autoreload 2
+
+# %%
+import os
+
+# Make sure we're running from the spam/ directory
+if os.path.basename(os.getcwd()) == "snorkel-tutorials":
+    os.chdir("spam")
+
+# %%
+from utils import load_spam_dataset
+
+df_train, _, df_valid, df_test = load_spam_dataset(delete_train_labels=False)
+
+# We pull out the label vectors for ease of use later
+Y_valid = df_valid["label"].values
+Y_train = df_train["label"].values
+Y_test = df_test["label"].values
+
+
+# %%
+df_train.head()
+
+# %% [markdown]
+# ## 2. Writing Transformation Functions
+#
+# Transformation Functions are functions that can be applied to a training example to create another valid training example. For example, for image classification problems, it is common to rotate or crop images in the training data to create new training inputs.
+#
+# Our task involves processing text. Some common ways to augment text includes replacing words with their synonyms, or replacing names entities with other entities. Applying these operations to a comment shouldn't change whether it is `SPAM` or not.
+#
+# Transformation functions in Snorkel are created with the `@transformation_function()` decorator, which wraps a function for taking a single data point and returning a transformed version of the data point.
+#
+# We start with a simple transformation function that changes a random character in the text to simulate a typo.
+
+# %%
+import string
+from snorkel.augmentation.tf import transformation_function
+
+
+@transformation_function()
+def change_character(x):
+    idx = np.random.choice(range(len(x.text) - 1))
+    char = np.random.choice(list(string.ascii_lowercase))
+    x.text = x.text[:idx] + char + x.text[idx + 1 :]
+    return x
+
+
+# %% [markdown]
+# ### Adding `pre` mappers.
+# Some TFs rely on fields that aren't present in the raw data, but can be derived from it.
+# We can enrich our data (providing more fields for the TFs to refer to) using map functions specified in the `pre` field of the transformation_function decorator (similar to `preprocessor` used for Labeling Functions).
+#
+# For example, we can use the fantastic NLP tool [spaCy](https://spacy.io/) to add lemmas, part-of-speech (pos) tags, etc. to each token.
+# Snorkel provides a prebuilt preprocessor for spaCy called `SpacyPreprocessor` which adds a new field to the
+# data point containing a [spaCy `Doc` object](https://spacy.io/api/doc).
+# For more info, see the [`SpacyPreprocessor` documentation](https://snorkel.readthedocs.io/en/master/source/snorkel.labeling.preprocess.html#snorkel.labeling.preprocess.nlp.SpacyPreprocessor).
```

Comment:
Note that it memoizes results, so it won't reparse text for each TF on a given example